tree-sitter/src/TreeSitter/GenerateSyntax.hs

```diff
@@ -50,60 +50,63 @@ astDeclarationsForLanguage language filePath = do
-    generatedDatatype name cons = DataD [] name [] Nothing cons deriveClause
+    generatedDatatype name cons typeParameterName = DataD [] name [PlainTV typeParameterName] Nothing cons deriveClause
 
 
 -- | Create TH-generated SymbolMatching instances for sums, products, leaves
-symbolMatchingInstance :: Ptr TS.Language -> Name -> String -> Q [Dec]
-symbolMatchingInstance language name str = do
+symbolMatchingInstance :: Ptr TS.Language -> Name -> String -> Name -> Q [Dec]
+symbolMatchingInstance language name str typeParameterName = do
   tsSymbol <- runIO $ withCString str (pure . TS.ts_language_symbol_for_name language)
   let tsSymbolType = toEnum $ TS.ts_language_symbol_type language tsSymbol
-  [d|instance TS.SymbolMatching $(conT name) where
+  [d|instance TS.SymbolMatching $(appT (conT name) (varT typeParameterName)) where
       showFailure _ node = "Expected " <> $(litE (stringL (show name))) <> " but got " <> show (TS.fromTSSymbol (nodeSymbol node) :: $(conT (mkName "Grammar.Grammar")))
       symbolMatch _ node = TS.fromTSSymbol (nodeSymbol node) == $(conE (mkName $ "Grammar." <> TS.symbolToName tsSymbolType str))|]
 
-symbolMatchingInstanceForSums ::  Ptr TS.Language -> Name -> [TreeSitter.Deserialize.Type] -> Q [Dec]
-symbolMatchingInstanceForSums _ name subtypes =
-  [d|instance TS.SymbolMatching $(conT name) where
+symbolMatchingInstanceForSums ::  Ptr TS.Language -> Name -> [TreeSitter.Deserialize.Type] -> Name -> Q [Dec]
+symbolMatchingInstanceForSums _ name subtypes typeParameterName =
+  [d|instance TS.SymbolMatching $(appT (conT name) (varT typeParameterName)) where
       showFailure _ node = "Expected " <> $(litE (stringL (show (map extractn subtypes)))) <> " but got " <> show (TS.fromTSSymbol (nodeSymbol node) :: $(conT (mkName "Grammar.Grammar")))
       symbolMatch _ = $(foldr1 mkOr (perMkType `map` subtypes)) |]
-  where perMkType (MkType (DatatypeName n) named) = [e|TS.symbolMatch (Proxy :: Proxy $(conT (toName named n))) |]
+  where perMkType (MkType (DatatypeName n) named) = [e|TS.symbolMatch (Proxy :: Proxy $(appT (conT (toName named n)) (varT typeParameterName))) |]
         mkOr lhs rhs = [e| (||) <$> $(lhs) <*> $(rhs) |]
         extractn (MkType (DatatypeName n) Named) = toCamelCase n
         extractn (MkType (DatatypeName n) Anonymous) = "Anonymous" <> toCamelCase n
 
 
 -- | Append string with constructor name (ex., @IfStatementStatement IfStatement@)
-constructorForSumChoice :: String -> TreeSitter.Deserialize.Type -> Q Con
-constructorForSumChoice str (MkType (DatatypeName n) named) = normalC (toName named (n ++ str)) [child]
-  where child = TH.bangType (TH.bang noSourceUnpackedness noSourceStrictness) (conT (toName named n))
+constructorForSumChoice :: String -> Name -> TreeSitter.Deserialize.Type -> Q Con
+constructorForSumChoice str typeParameterName (MkType (DatatypeName n) named) = normalC (toName named (n ++ str)) [child]
```

Comment:
This is called with a `traverse` above so you must order arguments according to the partial application you wish to see in the world.

Reply:
Wisdom.